packages/dart_frog/lib/src/body_parsers/form_data.dart

```diff
@@ -37,3 +54,145 @@ bool _isFormUrlEncoded(ContentType? contentType) {
+    final contentDisposition = part.headers['content-disposition'];
+    if (contentDisposition == null) continue;
+    if (!contentDisposition.startsWith('form-data;')) continue;
+
+    final values = _keyValueRegexp
+        .allMatches(contentDisposition)
+        .fold(<String, String>{}, (map, match) {
+      return map..[match.namedGroup('key')!] = match.namedGroup('value')!;
+    });
+
+    final name = values['name']!;
+    final fileName = values['filename'];
+
+    if (fileName != null) {
+      files[name] = UploadedFile(
+        fileName,
+        ContentType.parse(part.headers['content-type'] ?? 'text/plain'),
+        part,
+      );
+    } else {
+      final bytes = (await part.toList()).fold(<int>[], (p, e) => p..addAll(e));
+      fields[name] = utf8.decode(bytes);
+    }
+  }
+
+  return FormData(fields: fields, files: files);
+}
+
+/// {@template form_data}
+/// The fields and files of received form data request.
+/// {@endtemplate}
+class FormData with MapMixin<String, String> {
+  /// {@macro form_data}
+  const FormData({
+    required Map<String, String> fields,
+    required Map<String, UploadedFile> files,
+  })  : _fields = fields,
+        _files = files;
+
+  final Map<String, String> _fields;
+
+  final Map<String, UploadedFile> _files;
+
+  /// The fields that were submitted in the form.
+  Map<String, String> get fields => Map.unmodifiable(_fields);
+
+  /// The files that were uploaded in the form.
+  Map<String, UploadedFile> get files => Map.unmodifiable(_files);
+
+  @override
+  @Deprecated('Use `fields[key]` to retrieve values')
+  String? operator [](Object? key) => _fields[key] ?? _files[key]?.toString();
+
+  @override
+  @Deprecated('Use `fields.keys` to retrieve field keys')
+  Iterable<String> get keys => _fields.keys;
+
+  @override
+  @Deprecated('Use `fields.values` to retrieve field values')
+  Iterable<String> get values => _fields.values;
+
+  @override
+  @Deprecated(
+    'FormData should be immutable, in the future this will thrown an error',
+  )
+  void operator []=(String key, String value) => _fields[key] = value;
+
+  @override
+  @Deprecated(
+    'FormData should be immutable, in the future this will thrown an error',
+  )
+  void clear() => _fields.clear();
+
+  @override
+  @Deprecated(
+    'FormData should be immutable, in the future this will thrown an error',
+  )
+  String? remove(Object? key) => _fields.remove(key);
+}
+
+/// {@template uploaded_file}
+/// The uploaded file of a form data request.
+/// {@endtemplate}
+class UploadedFile {
+  /// {@macro uploaded_file}
+  const UploadedFile(
+    this.name,
+    this.contentType,
+    this._byteStream,
+  );
+
+  /// The name of the uploaded file.
+  final String name;
+
+  /// The type of the uploaded file.
+  final ContentType contentType;
+
+  final Stream<List<int>> _byteStream;
+
+  /// Returns the content of the file as a list of bytes.
+  ///
+  /// Can only be called once.
+  Future<List<int>> content() async {
```

Comment:
I'm still not a huge fan of this name (the distinction between bytes and content is not immediately clear imo) 🤔

Reply:
I am open to suggestions, this was the one I found quite hard to tackle. Can't find a good word that fits descriptively

Reply:
Was looking around for other examples in the wild and found [readAsBytes](https://api.flutter.dev/flutter/dart-io/File/readAsBytes.html) which has a very similar signature so wdyt about renaming `content` to `readAsBytes` and `bytes` to [`openRead`](https://api.flutter.dev/flutter/dart-io/File/openRead.html)?